### ISSUE TITLE
problem with tree farms due to hauler drones

### ISSUE THREAD

**Issue by Haze33E** [suppergerrie2/ModJam5#12]
**What happened?**
I was testing the drones and made a tree farm for large jungle trees

**What did you expect to happen?**
For it to produce lots to wood which it did. But with the limited item stack size pick-up and drop off the drones aren't able to keep up with the pace of the wood being produced. I started getting dozens of jungle wood blocks and saplings just laying on the ground and dropping FPS. Initially I had only 4 hauler drones picking up drops from the farm but they were bogged down very fast. I bumped it up to like 12 drones and even then they were still being outpaced by the production amount. Once the trees start growing they pump out wood really fast.

**How to reproduce?**
Arrange chests in a 2x2 pattern with jungle saplings they can be vertically staggered to achieve this. I haven't tried trapped chests.

I didn't test it but I'm guessing the same problem would arise from creating a large spruce tree farm. The drones don't have any issues with smaller and medium trees. But the jungle wood trees really showed how easily the drones can be bogged down by farm that produces lots of items.

![2018-07-13_20 27 08](https://user-images.githubusercontent.com/24802933/42718993-487ca13c-86dd-11e8-98c6-c81080aad05a.png)

**Comment by suppergerrie2** [comment]
I tested it with an inventory size of 16 and you still need quite a lot of drones, a big part of the problem is the AI (for example they sometimes wait for a log to drop from the leaves instead of picking it up). 
Atm there is a way to upgrade the storage size:
![image](https://user-images.githubusercontent.com/15769860/42723173-7eada446-8759-11e8-8966-36f67b22d3b6.png)


Another problem with the farm, but unrelated to this problem, is that the tree farm drone can suffocate in the trees.

**Comment by Haze33E** [comment]
I don't know the max upgrade levels or inventory sizes I haven't really looked into that but I had 16 level 8 storage hauler drones on both the spruce and jungle tree farms.  I think the problem mostly lies with how they're picking items up which is one at a time instead of picking them up how a player would with an automatic radius pick-up. I'm also guessing items also don't stack in their inventory because of the way I've noticed them picking up items. Which means wasted trips back and forth to a chest when they could just grab everything in a few stacks and go.

I think I may have had some drones get lost to the 2x2 trees it's hard to notice when there's so many of them. I'd find them in the deposit chests sometimes thinking a mob killed them I had archer and fighter drones running all over making the mobs mad. But maybe it was the big trees killing them and I just didn't notice.

I've been testing your mod alongside a couple others that pair pretty well with yours. At least when it comes to the drones depositing items into chests. **Faucets and Filters** and **Uppers** I was using them to make a filtered hopper/pipe system to replenish the saplings for the Tree farm drones. That's why my screenshot has dirt around the chests because I discovered the drones could get stuck on the pipes. I think it was the dark oak farm I made that stopped working and I noticed there was only 3 trees on each spot. Then I found the drone in a hole next to a pipe lol.
  
But yeah it's amazing the amount of wood coming out of those farms. If the hauler drones could grab all the drops it would be perfect. I haven't noticed any issues with the tree farm drones themselves though.

**Comment by suppergerrie2** [comment]
Atm there isn't a limit for the upgrade size. I'll add a limit and change the upgrade to be another forumula in the next update (right now it is default+upgrade level)

You can add filters to drones so you don't need a filter setup for your tree farm ;p

**Comment by Haze33E** [comment]
The filter was to push saplings into the chests that the tree farm drones use for home while the hauler drones collect the wood, apples, and saplings from trees. Otherwise they have to share the same chests and if I was to transport the items in the chests to a storage area the saplings would get sucked out with everything else. Also it keeps the Tree farm drone chest from filling up with wood/apples and them running out of saplings.  
  
As for my issues with the drone inventory from observing them to me it seems they grab 1 item, that item gets stored in 1 slot. If they grab an identical item it gets stored in a different slot even if that item is stackable. This is just my guess since I don't know how to mod otherwise I would have already taken a look at it and suggested a modification for you if it needed it. Also I'm really not a fan of them grabbing items 1 at a time or storing items in chests 1 at a time. I really wish they could grab items in stacks the way players can. If i dropped an entire stack of 64 items it should be able to grab that entire stack and store it within a single slot. I just want the drones to have a fluid and efficient method of grabbing and storing items. 

Again this was just my guess since I don't know how to mod if I assumed anything incorrectly I'm sorry. I'm just anxious to see this mod operating at full potential.

**Comment by suppergerrie2** [comment]
I fixed a bug which made the drone walk back to his home position instead of getting every item it could before walking back. This should probably help a little with this issue. This will be available in the next version

**Comment by Haze33E** [comment]
Hope to see this mod in 1.13

**Comment by suppergerrie2** [comment]
Don't expect it soon, the forge update will take a couple of months and only then I can start updating the mod.

**Comment by suppergerrie2** [comment]
Just another question, did you use any other mod that makes the leaves decay quicker? And you kept the randomtickspeed at the default right?

**Comment by Haze33E** [comment]
Everything was default except the mods I mentioned and your mod. I don't have any tree mods and I don't even know the commands for tick speed.

**Comment by suppergerrie2** [comment]
After some testing I had the following setup which worked good for a 2x2 jungle farm:
 - 4 tree farm drones, each with their own chest (By using chests and trapped chests).
 - 4 hauler drones with 3 storage upgrades and a filter for jungle saplings. Each one is placed on one of the tree farm chests.
 - 8 hauler drones with 5 storage upgrades and a filter for jungle logs and apples. They are placed in groups of 2 on 4 double chests.
 - I also used a speed boost for the hauler drones, this will be possible in the future.

Because I am gonna change the storage upgrade formula in the future you will probably need less storage upgrades. But I am gonna think about what is the best way between balance and fun.

**Comment by suppergerrie2** [comment]
https://www.youtube.com/watch?v=qfNASxMMmiQ&feature=youtu.be this would be the fully upgraded drone in the newest update. Maybe this is fast enough for a full tree farm?

**Comment by Haze33E** [comment]
Yeah that's pretty fast it still looks odd to me with the items over their head. If it just flashed for a second for each item and then faded away I think it would be better. That and being able to interact with a drone's inventory which would be good for crop farm drones and tree farm drones. So you can remove seeds and saplings from them so they swap to something else.

**Comment by suppergerrie2** [comment]
Can you test with the new update? The recipe for the upgrades are on the curseforge page

**Comment by Haze33E** [comment]
I already tested them when I was testing the crop farm drones. I used the same testing world from the tree farms I just had to chase down all the drones (which took a while) I had deployed to replace them all. When I updated the mod it erased all the upgrades on the drones I had previously deployed. I noticed they're faster then before and are better at picking up items. I only gave them the storage upgrade to level 6 cause I wasn't sure what the new limit is and I didn't even remember there was a speed upgrade.

**Comment by suppergerrie2** [comment]
Did the upgrades disappear from the items or from the drones?
The max level for storage is 15 and for speed is 10

**Comment by Haze33E** [comment]
The items and the drones both had their upgrades reset to 0 when I changed the mod to the newer version. I had a drone in my inventory and had to upgrade it again then chase down all the drones I had and replace them.

**Comment by Haze33E** [comment]
I think it might of been because I was using that debug version you sent me. I originally made that world with the mod from curseforge. Then I updated it with the debug version and then updated again when you updated the curseforge version. When I updated to the debug version you sent me I never replaced my drones or drone items. Guessing updating the mod a second time bugged them out and reset their upgrades.
